src/d.ts

```diff
@@ -36,6 +36,28 @@ export function isHNode(child: DNode): child is HNode {
 	return Boolean(child && (typeof child !== 'string') && child.type === HNODE);
 }
 
+/**
+ * Generic decorate function for DNodes, the modified in place based on the provided predicate
+ * and modifier functions.
+ *
+ * If no predicate is supplied then the modifier will be executed on all nodes.
+ */
+export function decorate(dNodes: DNode | DNode[], modifier: (dNode: DNode) => void, predicate?: (dNode: DNode) => void): DNode | DNode[] {
+	let nodes = Array.isArray(dNodes) ? [ ...dNodes ] : [ dNodes ];
+	while (nodes.length) {
+		const node = nodes.pop();
+		if (node) {
+			if (!predicate || predicate(node)) {
```

Comment:
The actual loop code originates from my original pr: https://github.com/dojo/widget-core/pull/270. The reason it's a while loop is I unrolled it from it being recursive. For the rest of the code I think the jsdoc comment is clear on how the functionality behaves:

```
* Generic decorator function for DNodes, the nodes are modified in place based on the provided predicate
 * and modifier functions.
 *
 * If no predicate is supplied then the modifier will be executed on all nodes.
```